File: scripts/phase4-green-path.mjs
#!/usr/bin/env node
/**
 * D14 — leave-simple Green path 联调
 *
 *   node scripts/phase4-green-path.mjs
 *   node scripts/phase4-green-path.mjs --pipeline-id 123   # 复用已有 pipeline
 *   node scripts/phase4-green-path.mjs --skip-artifacts     # 仅 API 断言，不查 workspace 目录
 *
 * 产出：.claude/evidence/phase4-d14-green-path.json
 *
 * 前置：start-dev.ps1（:5000 API 存活）
 */
import { login, pick } from './lib/jnpf-auth.mjs';
import {
  REPO_ROOT,
  assertGeneratedArtifacts,
  createPipeline,
  findSnapshot,
  getDeveloperStatus,
  getDiagnostics,
  getEvents,
  getSnapshots,
  isDeveloperGreen,
  log,
  parseSnapshotPayload,
  probeDeveloperApi,
  resolveGeneratedBackendRoot,
  runDeveloperOrchestrator,
  setupIr2Locked,
  waitDeveloperGreen,
  writeEvidence,
} from './lib/phase4-api.mjs';

const SKIP_ARTIFACTS = process.argv.includes('--skip-artifacts');
const PIPELINE_ARG = (() => {
  const idx = process.argv.indexOf('--pipeline-id');
  return idx >= 0 ? Number(process.argv[idx + 1]) : 0;
})();
const TIMEOUT_MS = Number(process.env.PHASE4_DEVELOPER_TIMEOUT_MS || 1_800_000);

const steps = [];

function step(name, pass, detail, extra = {}) {
  steps.push({ name, pass, detail, ...extra, at: new Date().toISOString() });
  log(pass ? 'PASS' : 'FAIL', name, detail);
  return pass;
}

async function main() {
  const session = await login();
  log('logged in as', session.account);

  let pipelineId = PIPELINE_ARG;
  if (!pipelineId) {
    pipelineId = await createPipeline(session, `P4-Green-${Date.now()}`);
    step('create-pipeline', true, `pipelineId=${pipelineId}`, { pipelineId });
  } else {
    step('reuse-pipeline', true, `pipelineId=${pipelineId}`, { pipelineId });
  }

  if (!PIPELINE_ARG) {
    try {
      await setupIr2Locked(session, pipelineId);
      step('ir2-locked', true, 'SystemDesignLocked + IR2_SystemDesign locked', { pipelineId });
    } catch (e) {
      step('ir2-locked', false, e.message, { pipelineId });
      throw e;
    }
  } else {
    const snaps = await getSnapshots(session, pipelineId);
    const system = findSnapshot(snaps, 'IR2_SystemDesign');
    const ok = pick(system, 'stabilityState', 'StabilityState') === 'locked';
    step('ir2-locked', ok, ok ? 'reused pipeline IR2 locked' : 'IR2_SystemDesign not locked', { pipelineId });
    if (!ok) process.exit(1);
  }

  await probeDeveloperApi(session, pipelineId);
  step('developer-api', true, 'developer/status reachable', { pipelineId });

  const alreadyGreen = PIPELINE_ARG > 0 && (await isDeveloperGreen(session, pipelineId));
  let green;
  if (alreadyGreen) {
    step('developer-run', true, 'skip re-run — pipeline already green', { pipelineId });
    green = { ok: true, types: (await getEvents(session, pipelineId)).map(e => pick(e, 'eventType', 'EventType')) };
  } else {
    const runId = await runDeveloperOrchestrator(session, pipelineId);
    step('developer-run', true, `orchestrator started runId=${runId}`, { pipelineId, runId });
    try {
      green = await waitDeveloperGreen(session, pipelineId, TIMEOUT_MS);
    } catch (e) {
      const types = e.lastEventTypes || [];
      step('developer-green', false, `${e.message}; events=${types.join(',')}`, { pipelineId, eventTypes: types });
      throw e;
    }
  }

  if (!green.ok) {
    step('developer-green', false, `failed: ${green.reason}`, { pipelineId, eventTypes: green.types });
    process.exit(1);
  }
  step('developer-green', true, 'CodeGeneratedStablePromoted + TestSuiteGenerated', {
    pipelineId,
    eventTypes: green.types,
  });

  const status = await getDeveloperStatus(session, pipelineId);
  const codegenStability = pick(status, 'codegenStability', 'CodegenStability');
  const sandboxPassed = pick(status, 'sandboxBuildPassed', 'SandboxBuildPassed') === true;
  step(
    'developer-status',
    alreadyGreen || (codegenStability === 'stable' && sandboxPassed),
    alreadyGreen
      ? `skip status — IR events already green (codegen=${codegenStability}, sandbox=${sandboxPassed})`
      : `codegenStability=${codegenStability}, sandboxBuildPassed=${sandboxPassed}`,
    { pipelineId, status },
  );

  const events = await getEvents(session, pipelineId);
  const eventTypes = events.map(e => pick(e, 'eventType', 'EventType'));
  const hasCodegenFailed = eventTypes.includes('CodegenFailed');
  const hasPromote = eventTypes.includes('CodeGeneratedStablePromoted');
  const archCritical = events.filter(
    e => pick(e, 'eventType', 'EventType') === 'ArchViolationDetected',
  ).length;
  step(
    'no-fail-events',
    !hasCodegenFailed || hasPromote,
    `CodegenFailed=${hasCodegenFailed}, promoted=${hasPromote}, ArchViolationDetected count=${archCritical}`,
    { pipelineId },
  );

  const snapshots = await getSnapshots(session, pipelineId);
  const codegenSnap = findSnapshot(snapshots, 'IR3_GeneratedCode');
  const testSnap = findSnapshot(snapshots, 'IR3_TestSuite');
  const codegenState = pick(codegenSnap, 'stabilityState', 'StabilityState');
  const testPayload = parseSnapshotPayload(pick(testSnap, 'payload', 'Payload'));
  const scenarioCount = Number(testPayload?.scenarioCount ?? 0);
  const scenarios = Array.isArray(testPayload?.scenarios) ? testPayload.scenarios : [];
  const testStable = pick(testSnap, 'stabilityState', 'StabilityState') === 'stable';
  step(
    'ir3-snapshots',
    alreadyGreen
      ? testStable && scenarioCount >= 3
      : codegenState === 'stable' && scenarioCount >= 3,
    alreadyGreen
      ? `reuse green — IR3_TestSuite=${pick(testSnap, 'stabilityState', 'StabilityState')}, scenarioCount=${scenarioCount}`
      : `IR3_GeneratedCode=${codegenState}, scenarioCount=${scenarioCount}`,
    { pipelineId, scenarioCount, scenarios: scenarios.map(s => s.caseId || s.CaseId) },
  );

  const diag = await getDiagnostics(session, pipelineId);
  const tenantId = pick(diag, 'tenantId', 'TenantId') || session.tenantId || '0';
  const projectId = String(pipelineId);
  const backendRoot = resolveGeneratedBackendRoot(tenantId, projectId);

  if (!SKIP_ARTIFACTS) {
    const artifacts = assertGeneratedArtifacts(backendRoot);
    step('workspace-artifacts', artifacts.pass, `${artifacts.detail} @ ${backendRoot}`, {
      pipelineId,
      tenantId,
      backendRoot,
      missing: artifacts.missing,
    });
  } else {
    step('workspace-artifacts', true, 'skipped (--skip-artifacts)', { pipelineId });
  }

  const allPass = steps.every(s => s.pass);
  const report = {
    phase: 'phase4-d14-green-path',
    pass: allPass,
    pipelineId,
    tenantId,
    projectId,
    repoRoot: REPO_ROOT,
    steps,
    eventTypes: [...new Set(eventTypes)],
    at: new Date().toISOString(),
  };

  const evidencePath = writeEvidence('phase4-d14-green-path.json', report);
  log('evidence →', evidencePath);
  log(allPass ? '[D14] PASS — leave-simple Green path' : '[D14] FAIL — see steps');

  if (!allPass) process.exit(1);
}

main().catch(err => {
  console.error('[D14] FATAL', err.message || err);
  process.exit(1);
});

File: scripts/phase4_green_path.py
#!/usr/bin/env python3
"""
D14 leave-simple Green path — Python 版（推荐）

  python scripts/phase4_green_path.py
  python scripts/phase4_green_path.py --pipeline-id 209
  python scripts/phase4_diagnose.py 209          # 失败时先诊断

依赖: pip install requests pycryptodome
产出: .claude/evidence/phase4-d14-green-path.json
"""
from __future__ import annotations

import argparse
import json
import os
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from lib.jnpf_client import (
    REPO_ROOT,
    diagnose,
    get_event_types,
    get_snapshots,
    is_ok,
    login,
    pick,
    probe_developer_api,
    run_developer,
    setup_ir2_locked,
    unwrap,
    wait_developer_green,
    write_evidence,
)
from jnpf_auth import api_request


def create_pipeline(name: str) -> int:
    req = (
        f"{name}：员工请假审批 leave-simple Green path，含 LeaveRequest 单表 MVP。"
        + "测" * 400
    )
    resp = api_request(
        "POST",
        "/api/studio/pipeline/execute/create",
        {"name": name, "userRequirement": req},
    )
    if not is_ok(resp):
        raise RuntimeError(f"create pipeline: {resp}")
    return int(pick(unwrap(resp), "pipelineId", "PipelineId"))


def assert_artifacts(tenant_id: str, project_id: str) -> tuple[bool, str]:
    backend = REPO_ROOT / "workspace" / "generated" / tenant_id / project_id / "backend"
    if not backend.exists():
        return False, f"missing {backend}"
    entity = list((backend / "Entitys").glob("*Entity.cs")) if (backend / "Entitys").exists() else []
    service = [
        p for p in (backend / "Services").glob("*Service.cs")
        if (backend / "Services").exists() and not p.name.endswith(".custom.cs")
    ] if (backend / "Services").exists() else []
    iface = list((backend / "Interfaces").glob("I*Service.cs")) if (backend / "Interfaces").exists() else []
    ok = len(entity) >= 1 and len(service) >= 1 and len(iface) >= 1
    return ok, f"entity={len(entity)} service={len(service)} iface={len(iface)} @ {backend}"


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--pipeline-id", type=int, default=0)
    parser.add_argument("--skip-artifacts", action="store_true")
    args = parser.parse_args()
    timeout_s = float(os.environ.get("PHASE4_DEVELOPER_TIMEOUT_MS", "1800000")) / 1000

    login()
    steps: list[dict] = []

    def step(name: str, ok: bool, detail: str, **extra):
        steps.append({"name": name, "pass": ok, "detail": detail, **extra, "at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())})
        print(f"[{'PASS' if ok else 'FAIL'}] {name}: {detail}")

    pipeline_id = args.pipeline_id or create_pipeline(f"P4-Green-Py-{int(time.time())}")
    step("pipeline", True, f"pipelineId={pipeline_id}", pipelineId=pipeline_id)

    if not args.pipeline_id:
        setup_ir2_locked(pipeline_id)
        step("ir2-locked", True, "SystemDesignLocked + skeleton stable")
    else:
        d = diagnose(pipeline_id)
        sk = next((s for s in d["snapshots"] if s["type"] == "IR0_Skeleton"), {})
        sys_ok = next((s for s in d["snapshots"] if s["type"] == "IR2_SystemDesign"), {})
        ok = sk.get("stability") == "stable" and sys_ok.get("stability") == "locked"
        step("ir2-locked", ok, f"skeleton={sk.get('stability')}, systemDesign={sys_ok.get('stability')}")
        if not ok:
            sys.exit(1)

    probe_developer_api(pipeline_id)
    step("developer-api", True, "developer/status OK")

    run_id = run_developer(pipeline_id)
    step("developer-run", True, f"runId={run_id}")

    try:
        green = wait_developer_green(pipeline_id, timeout_s)
    except TimeoutError as e:
        d = diagnose(pipeline_id)
        write_evidence("phase4-d14-green-path-fail.json", d)
        step("developer-green", False, str(e))
        print("\n失败诊断（最近 developer/tester 错误）:")
        for r in d["skillRuns"]:
            print(f"  {r['skillId']}: {r['error']}")
        print(f"\n运行: python scripts/phase4_diagnose.py {pipeline_id}")
        sys.exit(1)

    if not green.get("ok"):
        step("developer-green", False, green.get("reason", "unknown"))
        sys.exit(1)
    step("developer-green", True, "promote + TestSuiteGenerated")

    status = unwrap(api_request("GET", f"/api/studio/skills/developer/{pipeline_id}/status"))
    cs = pick(status, "codegenStability", "CodegenStability")
    sb = pick(status, "sandboxBuildPassed", "SandboxBuildPassed")
    step("developer-status", cs == "stable" and sb is True, f"codegenStability={cs}, sandbox={sb}")

    snaps = get_snapshots(pipeline_id)
    codegen = next((s for s in snaps if pick(s, "fragmentType", "FragmentType") == "IR3_GeneratedCode"), {})
    test = next((s for s in snaps if pick(s, "fragmentType", "FragmentType") == "IR3_TestSuite"), {})
    tp = test.get("payload") or test.get("Payload")
    if isinstance(tp, str):
        tp = json.loads(tp)
    count = int((tp or {}).get("scenarioCount") or 0)
    step(
        "ir3-snapshots",
        pick(codegen, "stabilityState", "StabilityState") == "stable" and count >= 3,
        f"GeneratedCode stable, scenarios={count}",
    )

    if not args.skip_artifacts:
        diag = unwrap(api_request("GET", f"/api/studio/ir/{pipeline_id}/diagnostics"))
        tenant = str(pick(diag, "tenantId", "TenantId") or "0")
        ok, detail = assert_artifacts(tenant, str(pipeline_id))
        step("workspace-artifacts", ok, detail)

    all_pass = all(s["pass"] for s in steps)
    report = {
        "phase": "phase4-d14-green-path",
        "tool": "python",
        "pass": all_pass,
        "pipelineId": pipeline_id,
        "steps": steps,
        "eventTypes": sorted(set(get_event_types(pipeline_id))),
    }
    path = write_evidence("phase4-d14-green-path.json", report)
    print(f"\n证据 → {path}")
    print("[D14]", "PASS" if all_pass else "FAIL")
    sys.exit(0 if all_pass else 1)


if __name__ == "__main__":
    main()
